### ISSUE TITLE
matplotlib does not build with numpy master

### ISSUE THREAD

**Issue by mdboom** [numpy/numpy#6551]
Something has changed in numpy.distutils that is causing matplotlib to not build.

I've tried `git bisecting`, but the history seems a little convoluted and it isn't helping to find the issue.

```
Traceback (most recent call last):
  File "/home/mdboom/.pyenv/versions/3.5.0/lib/python3.5/site-packages/setuptools/sandbox.py", line 152, in save_modules
    yield saved
  File "/home/mdboom/.pyenv/versions/3.5.0/lib/python3.5/site-packages/setuptools/sandbox.py", line 193, in setup_context
    yield
  File "/home/mdboom/.pyenv/versions/3.5.0/lib/python3.5/site-packages/setuptools/sandbox.py", line 237, in run_setup
    DirectorySandbox(setup_dir).run(runner)
  File "/home/mdboom/.pyenv/versions/3.5.0/lib/python3.5/site-packages/setuptools/sandbox.py", line 267, in run
    return func()
  File "/home/mdboom/.pyenv/versions/3.5.0/lib/python3.5/site-packages/setuptools/sandbox.py", line 236, in runner
    _execfile(setup_script, ns)
  File "/home/mdboom/.pyenv/versions/3.5.0/lib/python3.5/site-packages/setuptools/sandbox.py", line 46, in _execfile
    exec(code, globals, locals)
  File "/tmp/easy_install-ywbtwurq/numpy-1.10.1/setup.py", line 264, in <module>
    if isinstance(package, str):
  File "/tmp/easy_install-ywbtwurq/numpy-1.10.1/setup.py", line 256, in setup_package
    # Go through all of the packages and figure out which ones we are
  File "/home/mdboom/.pyenv/versions/3.5.0/lib/python3.5/site-packages/numpy/distutils/core.py", line 135, in setup
    config = configuration()
  File "/tmp/easy_install-ywbtwurq/numpy-1.10.1/setup.py", line 156, in configuration
    
  File "/home/mdboom/.pyenv/versions/3.5.0/lib/python3.5/site-packages/numpy/distutils/misc_util.py", line 1001, in add_subpackage
    caller_level = 2)
  File "/home/mdboom/.pyenv/versions/3.5.0/lib/python3.5/site-packages/numpy/distutils/misc_util.py", line 970, in get_subpackage
    caller_level = caller_level + 1)
  File "/home/mdboom/.pyenv/versions/3.5.0/lib/python3.5/site-packages/numpy/distutils/misc_util.py", line 907, in _get_configuration_from_setup_py
    config = setup_module.configuration(*args)
  File "numpy/setup.py", line 11, in configuration
    from setuptools.command.test import test as TestCommand
  File "/home/mdboom/.pyenv/versions/3.5.0/lib/python3.5/site-packages/numpy/distutils/misc_util.py", line 1001, in add_subpackage
    caller_level = 2)
  File "/home/mdboom/.pyenv/versions/3.5.0/lib/python3.5/site-packages/numpy/distutils/misc_util.py", line 970, in get_subpackage
    caller_level = caller_level + 1)
  File "/home/mdboom/.pyenv/versions/3.5.0/lib/python3.5/site-packages/numpy/distutils/misc_util.py", line 882, in _get_configuration_from_setup_py
    ('.py', 'U', 1))
  File "/home/mdboom/.pyenv/versions/3.5.0/lib/python3.5/imp.py", line 234, in load_module
    return load_source(name, filename, file)
  File "/home/mdboom/.pyenv/versions/3.5.0/lib/python3.5/imp.py", line 172, in load_source
    module = _load(spec)
  File "<frozen importlib._bootstrap>", line 693, in _load
  File "<frozen importlib._bootstrap>", line 673, in _load_unlocked
  File "<frozen importlib._bootstrap_external>", line 662, in exec_module
  File "<frozen importlib._bootstrap>", line 222, in _call_with_frames_removed
  File "numpy/core/setup.py", line 13, in <module>
    import sys
ImportError: No module named 'numpy._build_utils'

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "setup.py", line 381, in <module>
    **extra_args
  File "/home/mdboom/.pyenv/versions/3.5.0/lib/python3.5/distutils/core.py", line 108, in setup
    _setup_distribution = dist = klass(attrs)
  File "/home/mdboom/.pyenv/versions/3.5.0/lib/python3.5/site-packages/setuptools/dist.py", line 268, in __init__
    self.fetch_build_eggs(attrs['setup_requires'])
  File "/home/mdboom/.pyenv/versions/3.5.0/lib/python3.5/site-packages/setuptools/dist.py", line 313, in fetch_build_eggs
    replace_conflicting=True,
  File "/home/mdboom/.pyenv/versions/3.5.0/lib/python3.5/site-packages/pkg_resources/__init__.py", line 836, in resolve
    dist = best[req.key] = env.best_match(req, ws, installer)
  File "/home/mdboom/.pyenv/versions/3.5.0/lib/python3.5/site-packages/pkg_resources/__init__.py", line 1081, in best_match
    return self.obtain(req, installer)
  File "/home/mdboom/.pyenv/versions/3.5.0/lib/python3.5/site-packages/pkg_resources/__init__.py", line 1093, in obtain
    return installer(requirement)
  File "/home/mdboom/.pyenv/versions/3.5.0/lib/python3.5/site-packages/setuptools/dist.py", line 380, in fetch_build_egg
    return cmd.easy_install(req)
  File "/home/mdboom/.pyenv/versions/3.5.0/lib/python3.5/site-packages/setuptools/command/easy_install.py", line 629, in easy_install
    return self.install_item(spec, dist.location, tmpdir, deps)
  File "/home/mdboom/.pyenv/versions/3.5.0/lib/python3.5/site-packages/setuptools/command/easy_install.py", line 659, in install_item
    dists = self.install_eggs(spec, download, tmpdir)
  File "/home/mdboom/.pyenv/versions/3.5.0/lib/python3.5/site-packages/setuptools/command/easy_install.py", line 842, in install_eggs
    return self.build_and_install(setup_script, setup_base)
  File "/home/mdboom/.pyenv/versions/3.5.0/lib/python3.5/site-packages/setuptools/command/easy_install.py", line 1070, in build_and_install
    self.run_setup(setup_script, setup_base, args)
  File "/home/mdboom/.pyenv/versions/3.5.0/lib/python3.5/site-packages/setuptools/command/easy_install.py", line 1056, in run_setup
    run_setup(setup_script, args)
  File "/home/mdboom/.pyenv/versions/3.5.0/lib/python3.5/site-packages/setuptools/sandbox.py", line 240, in run_setup
    raise
  File "/home/mdboom/.pyenv/versions/3.5.0/lib/python3.5/contextlib.py", line 77, in __exit__
    self.gen.throw(type, value, traceback)
  File "/home/mdboom/.pyenv/versions/3.5.0/lib/python3.5/site-packages/setuptools/sandbox.py", line 193, in setup_context
    yield
  File "/home/mdboom/.pyenv/versions/3.5.0/lib/python3.5/contextlib.py", line 77, in __exit__
    self.gen.throw(type, value, traceback)
  File "/home/mdboom/.pyenv/versions/3.5.0/lib/python3.5/site-packages/setuptools/sandbox.py", line 164, in save_modules
    saved_exc.resume()
  File "/home/mdboom/.pyenv/versions/3.5.0/lib/python3.5/site-packages/setuptools/sandbox.py", line 139, in resume
    compat.reraise(type, exc, self._tb)
  File "/home/mdboom/.pyenv/versions/3.5.0/lib/python3.5/site-packages/setuptools/compat.py", line 65, in reraise
    raise value.with_traceback(tb)
  File "/home/mdboom/.pyenv/versions/3.5.0/lib/python3.5/site-packages/setuptools/sandbox.py", line 152, in save_modules
    yield saved
  File "/home/mdboom/.pyenv/versions/3.5.0/lib/python3.5/site-packages/setuptools/sandbox.py", line 193, in setup_context
    yield
  File "/home/mdboom/.pyenv/versions/3.5.0/lib/python3.5/site-packages/setuptools/sandbox.py", line 237, in run_setup
    DirectorySandbox(setup_dir).run(runner)
  File "/home/mdboom/.pyenv/versions/3.5.0/lib/python3.5/site-packages/setuptools/sandbox.py", line 267, in run
    return func()
  File "/home/mdboom/.pyenv/versions/3.5.0/lib/python3.5/site-packages/setuptools/sandbox.py", line 236, in runner
    _execfile(setup_script, ns)
  File "/home/mdboom/.pyenv/versions/3.5.0/lib/python3.5/site-packages/setuptools/sandbox.py", line 46, in _execfile
    exec(code, globals, locals)
  File "/tmp/easy_install-ywbtwurq/numpy-1.10.1/setup.py", line 264, in <module>
    if isinstance(package, str):
  File "/tmp/easy_install-ywbtwurq/numpy-1.10.1/setup.py", line 256, in setup_package
    # Go through all of the packages and figure out which ones we are
  File "/home/mdboom/.pyenv/versions/3.5.0/lib/python3.5/site-packages/numpy/distutils/core.py", line 135, in setup
    config = configuration()
  File "/tmp/easy_install-ywbtwurq/numpy-1.10.1/setup.py", line 156, in configuration
    
  File "/home/mdboom/.pyenv/versions/3.5.0/lib/python3.5/site-packages/numpy/distutils/misc_util.py", line 1001, in add_subpackage
    caller_level = 2)
  File "/home/mdboom/.pyenv/versions/3.5.0/lib/python3.5/site-packages/numpy/distutils/misc_util.py", line 970, in get_subpackage
    caller_level = caller_level + 1)
  File "/home/mdboom/.pyenv/versions/3.5.0/lib/python3.5/site-packages/numpy/distutils/misc_util.py", line 907, in _get_configuration_from_setup_py
    config = setup_module.configuration(*args)
  File "numpy/setup.py", line 11, in configuration
    from setuptools.command.test import test as TestCommand
  File "/home/mdboom/.pyenv/versions/3.5.0/lib/python3.5/site-packages/numpy/distutils/misc_util.py", line 1001, in add_subpackage
    caller_level = 2)
  File "/home/mdboom/.pyenv/versions/3.5.0/lib/python3.5/site-packages/numpy/distutils/misc_util.py", line 970, in get_subpackage
    caller_level = caller_level + 1)
  File "/home/mdboom/.pyenv/versions/3.5.0/lib/python3.5/site-packages/numpy/distutils/misc_util.py", line 882, in _get_configuration_from_setup_py
    ('.py', 'U', 1))
  File "/home/mdboom/.pyenv/versions/3.5.0/lib/python3.5/imp.py", line 234, in load_module
    return load_source(name, filename, file)
  File "/home/mdboom/.pyenv/versions/3.5.0/lib/python3.5/imp.py", line 172, in load_source
    module = _load(spec)
  File "<frozen importlib._bootstrap>", line 693, in _load
  File "<frozen importlib._bootstrap>", line 673, in _load_unlocked
  File "<frozen importlib._bootstrap_external>", line 662, in exec_module
  File "<frozen importlib._bootstrap>", line 222, in _call_with_frames_removed
  File "numpy/core/setup.py", line 13, in <module>
    import sys
ImportError: No module named 'numpy._build_utils'
```

**Comment by juliantaylor** [comment]
works fine for me with both git on python3.5, maybe its your environment

**Comment by mdboom** [comment]
What things in my environment should I be looking for?  All things being equal, if I roll back to the v1.10.1 tag in numpy, it works fine.

**Comment by mdboom** [comment]
Note that you have to clean the installation and build directories first to reproduce the bug.

**Comment by mdboom** [comment]
Ok -- so it does seem to be caused by some sort of environmental detritous.  Once I started an environment from scratch again, I'm no longer able to reproduce.

**Comment by mdboom** [comment]
Actually, I take that back.  The following are enough to reproduce:

```
pyenv install miniconda3-3.16.0
pyenv global miniconda3-3.16.0
conda install cython
cd numpy  # numpy master @ bf28b443
git checkout upstream/master
git clean -fxd
python setup.py install
cd ../matplotlib  # matplotlib master @ b6d18bbb
git checkout upstream/master
git clean -fxd
python setup.py install
```

... maybe it's conda-specific, so I'll try vanilla Python from source ...

```
pyenv install 3.5.0
pyenv global 3.5.0
pip install cython
cd numpy  # numpy master @ bf28b443
git checkout upstream/master
git clean -fxd
python setup.py install
cd ../matplotlib  # matplotlib master @ b6d18bbb
git checkout upstream/master
git clean -fxd
python setup.py install
```

And same for Python 2.7.10

However, if I pip install numpy, everything is happy:

```
pyenv install 3.5.0
pyenv global 3.5.0
pip install cython
pip install git+http://github.com/numpy/numpy#egg=numpy
cd ../matplotlib  # matplotlib master @ b6d18bbb
git checkout upstream/master
git clean -fxd
python setup.py install
```

In all cases, I've got the latest versions of pip and setuptools.

I'm going to see how the file trees differ between the two installation methods and report back.

**Comment by mdboom** [comment]
The difference is that the `pip` installation method installs an `.egg-info` directory for numpy, where as `setup.py install` installs an `.egg-info` file.  Replacing the file with the directory is enough to get the matplotlib build to work.  This difference alone isn't the whole story, though.  Numpy 1.10.1, which also works, installs an `.egg-info` file.

I'm officially out of my depth on Python packaging at this point, but hopefully that's helpful to track the issue down.

**Comment by charris** [comment]
Could you possibly do a bisect?

**Comment by mdboom** [comment]
I tried to bisect (see OP).  After the first step, it says:

```
The merge base a62abc814584047e22a630c92fe355783a5e9fc7 is bad.
This means the bug has been fixed between a62abc814584047e22a630c92fe355783a5e9fc7 and [f769c64026f5ff993ec1e28cbd9f39b8da75347d].
```

So I tried bisecting between a62a (bad) and f769 (good, also the head of 1.10.1)

```
Some good revs are not ancestor of the bad rev.
git bisect cannot work properly in this case.
Maybe you mistake good and bad revs?
```

Ok, so it doesn't like bad coming before good, let's try to reverse them (understanding this will be trying to find where on the 1.10.x branch the bug has already been fixed).  A few steps later and I get (remembering here it's opposite day and "bad" means "good"):

```
f769c64026f5ff993ec1e28cbd9f39b8da75347d is the first bad commit
commit f769c64026f5ff993ec1e28cbd9f39b8da75347d
Author: Charles R Harris <charlesr.harris@gmail.com>
Date:   Mon Oct 12 14:35:31 2015 +0000

    REL: Release 1.10.1

:100644 100644 01d1c04e6b3c3bbfd97bb3e350eee29b0c3dcaa4 698ae11d846a78b57276dace603a37e3f7a6a37e M      pavement.py
:100755 100755 491f346e6d9ccc7b44b1b68512712619502abda0 9edcb7dce05c7e08ba352c8d0b154846ab813285 M      setup.py
```

Odd.  This leads me to wonder if the fact of being a released version or not is the issue.  Indeed, if I modify numpy's setup.py on master to say `ISRELEASED = True`, voila, I have something that works.

Investigating further...

**Comment by mdboom** [comment]
Here's what I think is going on.

A dev version of numpy has the form, e.g. `1.11.0.dev0+bf29b44`.  That's a PEP440-compliant version.  All good so far.

However, it seems that when you do `setup.py install`, in installs the `.egg-info` file with the version `1.11.0.dev0_bf29b44` (having converted the `+` into a `_`).  This is not PEP440 compliant, and therefore compares with PEP440-compliant versions as if it were 0.0.  This is the version number -- in the file name, and not in any of the contents of the files -- that is used to determine if the version of numpy is new enough as specified in matplotlib's `install_requires` list.

When matplotlib's setup script runs, it thinks the installed numpy is too old, and therefore tries to fetch it from pip.  Then it does something that involves importing both parts of the newly downloaded numpy (in this case 1.10.1) and the installed numpy (in this case git master).  See the traceback in the OP, where parts of numpy are imported both from `/tmp` (which is the newly downloaded numpy) and `site-packages`.  This has worked for a very long time I presume by getting lucky.  In recent versions of numpy, something has changed such that this cross-importing of source and installed files doesn't work.

I think I need a packaging expert to help with this:  I think we need to either:

1. Prevent the version number from getting "sanitized" by setuptools when installing with `setup.py`

2. Maybe use some different incantation to specify matplotlib's dependencies that will work.  I seem to recall reading something about avoiding `install_requires` and `setup_requires` somewhere, but can't find it now.

In any case, I *doubt* this is a numpy bug -- really just setuptools confusion.  Feel free to close.

**Comment by charris** [comment]
Sounds like the problem is elsewhere. I'll leave this open for a while in case other folks run into the problem.

**Comment by mdboom** [comment]
Yeah -- it's in setuptools / how numpy and/or matplotlib uses setuptools.

**Comment by mdboom** [comment]
For reference: https://bitbucket.org/pypa/setuptools/issues/451

**Comment by cimarronm** [comment]
@mdboom, Looks like you have been having a whole bunch of fun ;-). I try to avoid `setuptools` like the plague.

**Comment by njsmith** [comment]
I wonder if it's time to start forcing people to run 'pip install' instead
of calling setup.py directly... That's the long term future, and these
kinds of problems are a ridiculous waste of everyone's time, and basically
insoluble otherwise. (See also: the bug reports we get every release or two
because 'setup.py install' leaves you with a package that can't be cleanly
uninstalled, and people trying to upgrade end up with a mix of files from
different numpy versions. pip fixes that.)
.
This would break setup_requires, because those don't go through pip -- if
you use setup_requires, then your setup.py will just go ahead and import
the easy_install code and start easy_installing away. (Yes, really...) But
from this bug report it sounds like setup_requires is broken for numpy
already (aside from all the many reasons that setup_requires is broken in
general).
.
I guess the people who would be most affected by this are packagers like
Debian / fedora / conda.
.
@juliantaylor: how horrible would this be for Debian?
.
(Anyone know who we should cc for other package systems?)

Sounds like the problem is elsewhere. I'll leave this open for a while in
case other folks run into the problem.

—
Reply to this email directly or view it on GitHub
<https://github.com/numpy/numpy/issues/6551#issuecomment-150379576>.

**Comment by mdboom** [comment]
I'm all for pushing people toward pip -- is there any magic incantation to push people toward that when using `setup.py` interactively (as opposed to within `pip` itself where I assume it still must work).

**Comment by kyleabeauchamp** [comment]
How would this work with conda?

**Comment by njsmith** [comment]
@mdboom: I think something something like this would work (untested):
```
# at the *top* of setup.py:
import sys
if "setuptools" not in sys.modules:
    # If pip were doing its magic, setuptools would have been imported before we started.
    # It hasn't been, so we can't be running under pip.
    # We could do anything in response, but for example...
    import os
    if "install" in sys.argv and "NUMPY_SETUP_PY_FORCE" not in os.environ:
        sys.stderr.write("To install numpy, use 'pip install numpy' or 'pip install <path to numpy checkout.\n")
        sys.stderr.write("Calling setup.py directly or via easy_install is not supported.\n")
        sys.stderr.write("To proceed at your own risk, set the environment variable:\n")
        sys.stderr.write("    NUMPY_SETUP_PY_FORCE=1\n")
```

@kyleabeauchamp: For installing numpy from source into a conda environment as an end user, nothing much would change -- you'd have to use `pip install .` instead of `python ./setup.py install`, and it would work as well as it ever does. For building a conda package of numpy... well, we don't know if it would cause any problems, because the conda recipe for numpy is proprietary :-). But probably they're currently calling `python setup.py install` internally, and would need to switch this to call `pip install .` instead. (Or, if lazy and we do something like the code above, to switch this to call `env NUMPY_SETUP_PY_FORCE=1 python setup.py install`.)

**Comment by kyleabeauchamp** [comment]
PS: the continuum recipe may be proprietary, but there's definitely an open source version that works for openblas.  But you're right that it may presumably be as easy as requiring pip as a conda dependency and changing the build script to use `pip install` rather than `python setup.py install`.

https://github.com/conda/conda-recipes/tree/master/numpy-openblas

**Comment by njsmith** [comment]
Ah, interesting!

If you have a moment to test, it would be an interesting data point to find out whether it works to add pip as a build-dependency and make the build line `$PYTHON -m pip install .`

**Comment by kyleabeauchamp** [comment]
Yes, the pip-based conda recipe seems to build and install, although I didn't spend too much time poking at the resulting package.

**Comment by rgommers** [comment]
This is a duplicate of gh-6257. The problem in this case is *not* using setuptools. @mdboom if you use ``python setupegg.py install`` (which injects ``setuptools``) instead of ``python setup.py install`` to install numpy master then I bet it works.

**Comment by rgommers** [comment]
@njsmith I think you're forgetting all the other commands you can use with ``python setup.py``. How would you run ``egg_info`` through pip for example? Maybe only erroring on ``python setup.py install`` could be acceptable, although we're going to get a lot of complaints about that from users I suspect.

**Comment by njsmith** [comment]
@rgommers: notice the `if "install" in sys.argv` in my sample code :-). (That said, I really can't think of many use cases for running setup.py directly -- am I missing some? I know weirdos like us sometimes want to, say, manually run `egg_info` and then manually run `build_ext -i` and then manually fiddle with our PYTHONPATH instead of running `pip install --editable .` like normal people, but... I guess we do still use `bdist_win`, but other than that?)

**Comment by juliantaylor** [comment]
I exclusively use setup.py install, I honestly can probably count the times I used pip in my live on 2 hands ...
but then I'm probably not an average user, uninstall for me is just rm -rf <prefix>

ralfs point about setup.py commands is a good one, can you execute all of them from pip? (build_ext --inplace, --parallel, cython, test, ...)

I think a saner approach would be to make setuptools a requirement for numpy installation, its needed by pip and newer pythons have this ensurepip module to pull it.

**Comment by mdboom** [comment]
> I think a saner approach would be to make setuptools a requirement for numpy installation, its needed by pip and newer pythons have this ensurepip module to pull it.

@juliantaylor: In matplotlib, we combined our `setup.py` and `setupegg.py` years ago -- so `setup.py` always uses setuptools, and we haven't had any complaints (at least about that specifically).  So in that sense I agree that that might be a good lower-risk option for numpy for now.  

I'm still a bit puzzled by why setuptools even gets imported by numpy's vanilla `setup.py`, though (see traceback at top of this issue).  But I can confirm that using `setupegg.py` rather than `setup.py` resolves the issue.

**Comment by rgommers** [comment]
> I'm still a bit puzzled by why setuptools even gets imported by numpy's vanilla setup.py, though (see traceback at top of this issue). But I can confirm that using setupegg.py rather than setup.py resolves the issue.

Because you're not invoking the numpy build directly but via ``easy_install``.

The long version: the ``setup()`` call at the end of MPL's ``setup.py`` is not from distutils as the command above claims but from setuptools. There's some quite hard to follow parsing of MPL dependencies, but it ends with numpy in ``setup_requires`` and ``install_requires``. Passing ``*_requires=='numpy'`` to setuptools invokes ``easy_install`` to go install those dependencies.

**Comment by mdboom** [comment]
Thanks @rgommers.  That makes sense.

**Comment by rgommers** [comment]
> notice the if "install" in sys.argv in my sample code :-).

I did now:) Reading too fast before running out the door ....

> I guess we do still use bdist_win, but other than that?

How about installing with a non-default compiler? On Windows you either need to go the non-free MSVC (and ifort) route, or do something like:

    python setup.py config --compiler=mingw32 build --compiler=mingw32 install

And yes I know you can put a magic file (forgot what the name was) somewhere to let distutils know that you *always* want to use mingw, but that may not always be appropriate.

**Comment by njsmith** [comment]
> ralfs point about setup.py commands is a good one, can you execute all of them from pip? (build_ext --inplace, --parallel, cython, test, ...)

Again, no one is arguing for disabling setup.py entirely, just setup.py install :-). And even that we'd leave accessible to experts. To answer the question though: sorta (`install -e`), yes (awkwardly via `--install-option`), no, no.

> I think a saner approach would be to make setuptools a requirement for numpy installation, its needed by pip and newer pythons have this ensurepip module to pull it.

Setuptools' raw install command fixes some-but-not-all of distutils brokenness (at least it won't mangle the version number, but users will still be bitten by broken uninstall/upgrade support), but also adds more brokenness of its own that pip will paper over. Ralf could probably give you a longer list, but the mess you get if you don't use --single-version-externally-managed is pretty compelling to me.

> Passing `*_requires=='numpy'` to setuptools invokes easy_install to go install those dependencies.

Small correction: what you say is always true (right now) for setup_requires, and it's probably true for install_requires if you're using setup.py install directly with setuptools, but if you're using pip then install_requires will be handled by pip. (This is why pip does the awkward dance of first running egg_info before actually building anything: it's peeking at the install_requires metadata.)

**Comment by njsmith** [comment]
> EDIT: and we get bug reports from people trying setup.py clean, etc.

Which is a buggy piece of junk, weren't you saying recently? :-) Maybe this should error out with some instructions on how to use git, regardless ;-)

> python setup.py config --compiler=mingw32 build --compiler=mingw32 install

Great example, thank you! IIRC you can give pip options to pass to `build`, but it doesn't know anything about `config`, so that doesn't help.

Some possibilities to think about:

We could teach config and build to check a config file, or config to look at build's options. Of course this requires touching numpy.distutils so maybe not.

This might work:
```
python setup.py config --compiler=mingw32 build --compiler=mingw32
pip install .
```

This would work as well as it does currently:
```
set NUMPY_SETUP_PY_FORCE=1
python setup.py config --compiler=mingw32 build --compiler=mingw32 install
```

This would work better:
```
python setup.py config --compiler=mingw32 build --compiler=mingw32 bdist_wheel
pip install dist/*.whl
```

Obviously we can't actually make any decisions here without consulting the mailing list, but it's useful to get some first impressions and get a sense of what kind of proposal might even make sense or not.

**Comment by rgommers** [comment]
> Which is a buggy piece of junk, weren't you saying recently? :-) Maybe this should error out with some instructions on how to use git, regardless ;-)

It is a buggy piece of junk:) However the report came from Antoine Pitrou (and not a user who simply doesn't know better), who had a workflow that basically made it impossible to use ``git clean -xdf``.

**Comment by rgommers** [comment]
@njsmith how about something simpler and not requiring 2 commands: if the exact install command was one of 

- ``python setup.py install``
- ``python setup.py install --user``

then error out, and otherwise assume that the user knows what he's doing.

**Comment by njsmith** [comment]
> However the report came from Antoine Pitrou (and not a user who simply doesn't know better), who had a workflow that basically made it impossible to use `git clean -xdf`.

Are you referring to #5467? There's nothing in there saying that `git clean -xdf` was difficult or impossible, but maybe there's context I'm missing. Let's ask :-) @pitrou: would `setup.py clean` issuing an error that just says "this doesn't work, use git clean -xdf instead" have helped your situation, made it worse, or...?

> how about something simpler and not requiring 2 commands: if the exact install command was one of [...] then error out, and otherwise assume that the user knows what [they're] doing.

I'm not sure that users do know what they're doing when it comes to `setup.py install`. (I'm still not even certain that I understand all the nuances). And if the goal is to reduce confusion around weird edge cases then maybe we don't want `python setup.py install -j4` to silently produce a quasi-broken install :-). And we probably need some override anyway to allow people to do a simple `python setup.py install` if they really really want to. So my preference would still be something explicit rather than implicit. (If we want to keep it to a single command, then maybe instead of an envvar we could use a command-line option, `--force` or `--i-know-what-im-doing` or whatever?)

But I do think I like your proposal better than the status quo. Anything that lets us move away from supporting like 3+ different install methods that each have their own arcane bugs. (Plus nudging people to stop calling `setup.py` directly helps lay the groundwork for a world where we don't use distutils/setuptools at all.)

**Comment by rgommers** [comment]
I like ``--force`` a bit better. Anyway, I think we're thinking in the same direction - the difference is in the details.

**Comment by rgommers** [comment]
> Are you referring to #5467? 

IIRC there was a longer discussion where @pitrou described his workflow (keeping personal test scripts inside the git repo), but I can't find it right now.

**Comment by pitrou** [comment]
>  would setup.py clean issuing an error that just says "this doesn't work, use git clean -xdf instead" have helped your situation, made it worse, or...?

It would certainly have helped since it would have given me the next step to unlock the situation. It probably wouldn't help people who run `setup.py clean` from a tarball rather than a git checkout, but I don't think that happens often.

**Comment by pitrou** [comment]
By the way, it seems it should be `git clean -Xdf` not `git clean -xdf` (note capital X): the former only cleans ignored files, while the latter cleans everything.

**Comment by pitrou** [comment]
> keeping personal test scripts inside the git repo

Surely you mean "git working copy", not "git repo". That's what I do when experimenting or trying to reproduce bugs, yes.

**Comment by rgommers** [comment]
Thanks @pitrou. So let's raise an error with a clear message then.

> Surely you mean "git working copy", not "git repo". That's what I do when experimenting or trying to reproduce bugs, yes.

That's a very svn-mindset-like comment:) You do have a full repository on your own hard drive, not just a working copy. I think the terminology is right.

> By the way, it seems it should be git clean -Xdf not git clean -xdf (note capital X): the former only cleans ignored files, while the latter cleans everything.

That's the wrong way around, lower-case ``-x`` cleans it all:

    $ echo "random text" > notignored.txt
    $ git clean -xdf
    Removing notignored.txt

**Comment by rgommers** [comment]
So maybe we should make a longer list of commands that should raise clear error messages. For example ``python setup.py test`` should raise an error saying to use ``runtests.py`` or ``pip install`` followed by ``import numpy as np; np.test()``.

**Comment by pitrou** [comment]
@rgommers, it seems like you are misunderstanding me.

> You do have a full repository on your own hard drive, not just a working copy. I think the terminology is right

No. If I was "keeping personal scripts in the repo", that would mean I would version them in git. I don't. I just keep them as files at the base of the working copy.

(note "working copy" may sound svn-ish, but git man pages use the very similar "working tree" and mercurial help pages seem to like "working directory")

> That's the wrong way around, lower-case -x cleans it all:

And precisely, that's what I don't want. I only want to clean files produced by *building numpy*, not my own files which happen to be lying in the directory.

**Comment by rgommers** [comment]
@pitrou okay I get what you mean now. Then I guess we can document both ``-X`` and ``-x`` with an explanation of the difference. The latter should still be preferred for most users. Note that there's a lot in ``.gitignore`` that will still be cleaned up even with ``-X``:

- ``site.cfg``
- ``*.zip``, ``*.tar``, etc.
- log files, sql(ite) databases, etc.
- patch/diff files

Basically only ``*.py`` and ``*.c`` are guaranteed not to be in there.....

**Comment by njsmith** [comment]
Proposal to disable `setup.py install` sent to list: https://mail.scipy.org/pipermail/numpy-discussion/2015-October/074086.html

**Comment by NelleV** [comment]
My 2 cents is that you can't expect people to stop using setup.py install and use pip install (one is the default way to install python packages, while the other uses a package that isn't even installed on many OS). And you can't expect people to read the documentation for something as simple as installing a package (I sure never do.).
Wouldn't an option be to fix the bug?

**Comment by rgommers** [comment]
@NelleV thanks for your 2c. We've already come to the conclusion I think that in the short term that's going to be too large a change. Also note that the bug in this issue isn't directly related to the ``setup.py install`` issue. 

The fix for this issue is ``import setuptools``. We've managed to avoid that for a long time, but unfortunately it looks like we have to now......

**Comment by njsmith** [comment]
> My 2 cents is that you can't expect people to stop using setup.py install and use pip install (one is the default way to install python packages, while the other uses a package that isn't even installed on many OS).

FYI -- the long-term plan in the upstream python packaging world (regardless of what we do here) is to convince everyone to stop running setup.py install directly and make pip install the default way to install python packages from source. If your Python doesn't have pip then you should file a bug with whoever you got Python from, because it's [officially included with modern Python interpreter releases](https://pip.readthedocs.org/en/stable/installing/#pip-included-with-python); not providing pip is a bug on par with leaving out part of the standard library. (Or if you're stuck on an old version of Python for some reason then installing pip is still completely trivial -- certainly much easier than building numpy from source!)

> And you can't expect people to read the documentation for something as simple as installing a package (I sure never do.).

That's why the "documentation" would be printed when people tried running `setup.py install` :-).

> Wouldn't an option be to fix the bug?

As Ralf says, this particular bug can be fixed by using setuptools, but using setuptools brings along other obnoxious behavior with it, and there would still remain various bugs that couldn't be fixed at all except through using `pip install`. (In particular, it's impossible to reliably uninstall or upgrade a package that was installed via `setup.py` -- the only reliable way to get rid of it is to nuke your whole python environment and start over. If you try to upgrade in place then it will pretend to work, but sometimes it will leave you with a corrupted install and weird problems that will confuse maintainers to no end. Isn't software wonderful!)

**Comment by jakirkham** [comment]
Saw mention of building with `conda`. This may not have come up yet, but `conda` makes sure any environment it creates that has `python` it will include `pip` and `setuptools` (amongst other things). Any time a recipe is built this happens too.
